Implement personal rooms

Joimer opened the pull request:
Personal rooms are temporary rooms that are neither stored nor logged.
The purpose of this kind of room is twofold:
First, it serves as a multiple user Private Message conversation.
Second, it serves as a private friend chat.
This eliminates the need to register private rooms for groups of friends.

Personal rooms are destroyed when no one is on them.
They are to be used with /invite only.
They can be created by anyone.
They require a name, but the creator name will be present on title.

kotarou3 commented on commands.js:
```diff
@@ -286,6 +286,41 @@ var commands = exports.commands = {
+
+		// Create the name of the room using user's ID and the chosen name.
+		var roomName = target + " (" + user.userid + "'s room)";
+		id = toId(roomName);
```
`toRoomId` (or was it `toRoomid`?) here

Joimer replied:
``makechatroom`` uses ``toId``

kotarou3 replied:
@Zarel pls

Zarel replied:
`-` has special meaning in room IDs. No user should be able to add `-` to a room ID.

panpawn commented:
I think that this is a great feature, awesome work @Joimer

kotarou3 commented:
Basically IRC rooms for those wondering

Slayer95 commented on rooms.js:
```diff
@@ -1676,6 +1679,16 @@ var ChatRoom = (function () {
+
+		// If it's a personal room, it gets destroyed when there are 0 users on it.
+		// We don't run ChatRoom.prototype.destroy because this room ain't registered.
+		// We have to clean up here.
```
``room.destroy()`` should work for unregistered rooms...

Zarel replied:
`room.destroy()` is the recommended way to destroy any room.

xfix commented on rooms.js:
```diff
@@ -1676,6 +1679,16 @@ var ChatRoom = (function () {
+		// We don't run ChatRoom.prototype.destroy because this room ain't registered.
+		// We have to clean up here.
+		if (this.isPersonal && this.userCount === 0) {
+			this.users = null;
```
Any reason why explicitly setting ``users`` to ``null`` is needed?

ascriptmaster commented on commands.js:
```diff
@@ -286,6 +286,46 @@ var commands = exports.commands = {
+		id = toId(roomName);
+
+		// Personal rooms can always be created, but the name must still be unique.
+		if (Rooms.search(id)) return this.sendReply("You already have created room '" + roomName + "'.");
```
Just say it already exists. This prevents strange shenanigans if a server admin creates a room called "This is Joim's room" and then Joim tries to make a personal room called "This is".

Zarel commented:
We should not allow users to create rooms in the global namespace. Besides what happens if someone makes a room named `teambuilder`...

My original sketch of this feature was that IDs should begin with `groupchat-`. Also, they'd expire after, say, an hour of inactivity (like battles).

Joimer commented:
The idea is that their name always get attached to the room, so it makes for unique room names. I thought it wasn't necessary to create a namespace for it.

``groupchat-`` is a nice idea, though. About expiring, I think that if everyone is gone, the room can be safely destroyed, don't you think so? After all, unlike battles, if they are all gone, it means there's nothing to see for the group of friends.

Zarel commented:
Is that how the feature's implemented? That works for me, I guess.

Joimer commented:
Yeah, line 300:
``var roomName = target + " (" + user.userid + "'s room)";``
Seemed good to me, as it has the name, like in a Smogon conversation, and also indicates who created the room.

Zarel commented:
I was asking about

> About expiring, I think that if everyone is gone, the room can be safely destroyed, don't you think so?

is that how it's currently implemented?

As for userid in room name, that's not enough uniqueness, especially since you're not using the namespace separator `-`

For instance, roomid `foobarbazsroom` - is this `foo (barbaz's room)` or `foobar (baz's room)`?

Joimer commented:
Yeah, I manage it in ``ChatRoom.prototype.onLeave``, if the room is empty it gets immediately destroyed.

As for the room id, you're right on that, I'll use ``-`` instead.

Zarel commented:
I think `groupchat-` might be enough. Maybe `groupchat-USERID-`, but that's probably unnecessary.

I also think people should be able to create rooms without a name to default to a random 6-digit number (so enough that it won't be guessed).

Joimer commented:
I've decided in favour of ``groupchat-userid-`` because that means different groups can have the same room name without having to add the long "(name's room)". string. Also added some checks on ``writeRoomData`` places, to avoid any possible problem in any possible server.

Zarel commented:
I'm not sure it's a good thing to allow different groups to have the same room name.

Joimer commented:
Okay, updated the PR.

Zarel commented:
Now time to discuss aliases! I was planning to call this feature "group chat" to emphasize that it's intended to be an extension of the PM feature for >2 people.

So I think the primary command should be something like `/groupchat`.

Maybe have it be a family of commands. `/groupchat new` to make a groupchat, `/groupchat` to list group chats and give an overview of commands for making more.

Zarel commented on commands.js:
```diff
@@ -509,6 +558,8 @@ var commands = exports.commands = {
 
 		if (!this.can('makeroom')) return false;
 
+		if (room.isPersonal) return this.errorReply("Personal rooms can only have one room owner.");
```
I wonder if this should be necessary.

Zarel commented on commands.js:
```diff
@@ -438,6 +485,7 @@ var commands = exports.commands = {
 		}
 		target = target.trim();
 		if (!this.can('declare', null, room)) return false;
+		if (room.isPersonal) return this.errorReply("Personal rooms configuration can't be changed.");
```
I wonder if this is necessary?

Zarel commented on commands.js:
```diff
@@ -376,6 +421,7 @@ var commands = exports.commands = {
 	officialchatroom: 'officialroom',
 	officialroom: function (target, room, user) {
 		if (!this.can('makeroom')) return;
+		if (room.isPersonal) return this.errorReply("Personal rooms configuration can't be changed.");
```
This requires the `makeroom` permission anyway, so I'm not sure this is necessary.

Zarel commented on commands.js:
```diff
@@ -346,6 +390,7 @@ var commands = exports.commands = {
 
 	modjoin: function (target, room, user) {
 		if (!this.can('privateroom', null, room)) return;
+		if (room.isPersonal) return this.errorReply("Personal rooms configuration can't be changed.");
```
`modjoin` should definitely be open to personal rooms. Basically any feature open to battle rooms should be open to personal rooms.

Zarel commented on commands.js:
```diff
@@ -610,7 +662,7 @@ var commands = exports.commands = {
 
 		if (targetUser) targetUser.updateIdentity(room.id);
-		if (room.chatRoomData) Rooms.global.writeChatRoomData();
+		if (room.chatRoomData && !room.isPersonal) Rooms.global.writeChatRoomData();
```
In theory, personal rooms shouldn't have `chatRoomData`...

Joimer replied:
``chatRoomData`` is created on the ``ChatRoom`` constructor, but yeah, can be changed so it doesn't if necessary.

Zarel replied:
Yes, that should be changed.

Zarel commented on commands.js:
```diff
@@ -1243,7 +1296,7 @@ var commands = exports.commands = {
 		this.logModCommand(user.name + " set modchat to " + room.modchat);
 
-		if (room.chatRoomData) {
+		if (room.chatRoomData && !room.isPersonal) {
```
This _definitely_ shouldn't have ` && !room.isPersonal`

Zarel commented on rooms.js:
```diff
@@ -650,15 +650,17 @@ var GlobalRoom = (function () {
 		return this;
 	};
-	GlobalRoom.prototype.addChatRoom = function (title) {
+	GlobalRoom.prototype.addChatRoom = function (title, personal) {
```
I would call this argument `isPersonal`.

Zarel commented on commands.js:
```diff
@@ -820,6 +872,7 @@ var commands = exports.commands = {
 		if (targetRoom.isPrivate) {
 			return this.errorReply("Do not redirect users to private rooms. Use /invite if you must, but honestly, the room probably wants to be kept private.");
 		}
+		if (targetRoom.isPersonal) return this.sendReply("Users cannot be redirected to a personal room.");
```
I don't think this is necessary... I wouldn't call it wrong, either, but more code complexity is bad.

Zarel commented on commands.js:
```diff
@@ -534,6 +585,7 @@ var commands = exports.commands = {
 
 		if (room.auth[userid] !== '#') return this.sendReply("User '" + name + "' is not a room owner.");
 		if (!this.can('makeroom')) return false;
+		if (room.isPersonal) return this.errorReply("Personal rooms ownership can't be changed.");
```
I'm not seeing a strong reason why not.

Joimer commented:
I've added an expire timer that will destroy a personal room that hasn't been updated in 40 minutes, just like battles. I feel this is necessary to keep rooms clean.

Zarel commented on commands.js:
```diff
@@ -286,6 +287,50 @@ var commands = exports.commands = {
+	personalroom: 'groupchat',
+	groupchat: function (target, room, user) {
+		// First and foremost, check the number of personal rooms owned by the user.
+		if (user.personalRooms && user.personalRooms >= MAX_PERSONAL_ROOMS) return this.sendReply("You can't own more than " + MAX_PERSONAL_ROOMS + " personal rooms.");
```
The way this is written, this is very easy to get around. I'd suggest just using ResouceMonitor, which uses a more DoS-proof method of limiting resource usage.

Zarel replied:
Specifically, by doing it by IP instead of by user object.

Joimer replied:
Yeah, I was thinking to have both a hard limit per user and a rate limit per IP per room creation.

Zarel commented on commands.js:
```diff
@@ -286,6 +287,50 @@ var commands = exports.commands = {
+		var id = toId(target);
+		// If there isn't any, we roll our own random number title.
+		if (!id) {
+			target = Math.floor(Math.random() * 10000);
```
I said six digits, this is four digits which is not enough entropy. :(

Zarel replied:
Even six digits is a bit low; we should probably do 8. Let's do 8.

Zarel commented on commands.js:
```diff
@@ -286,6 +287,50 @@ var commands = exports.commands = {
+		if (Rooms.global.addChatRoom(roomName, true)) {
+			var targetRoom = Rooms.search(roomName);
+			// Make the room private, as it can't be public.
+			targetRoom.isPrivate = true;
```
Should be hidden by default, not private.

Zarel replied:
Maybe even public by default. They obviously won't be shown in the room list, but they could be shown on the user menu and /whois.

panpawn replied:
Would they be shown on /whois for regular users or just global staff?

Zarel commented on commands.js:
```diff
@@ -286,6 +287,50 @@ var commands = exports.commands = {
+			targetRoom.isPrivate = true;
+			targetRoom.chatRoomData.isPrivate = true;
+			// The room is modjoin by default. This shouldn't be changed.
+			targetRoom.modjoin = true;
```
Why?

Also, new plan: `/invite` automatically promotes to the modjoin rank when modjoin is enabled.

Zarel replied:
`modjoin = true` is also meaningless unless you have modchat on.

Did you mean `modjoin = '+'`? Or did you mean `modjoin = true; modchat = '+'`?

Joimer replied:
This and making it private (or hidden now) by default is because it's treated like a Conversation in Smogon or just a group Private Message, that is, cannot be seen by people unless they are invited.

Zarel replied:
Hm, if it's to be treated like a group PM, then drivers+ probably shouldn't be able to see/moderate it (i.e. it should be private instead of hidden)...

Joimer commented:
I've added ResourceMonitor time rate rate and increased entropy in random group chat titles.

Zarel commented on commands.js:
```diff
@@ -306,6 +351,7 @@ var commands = exports.commands = {
 	hiddenroom: 'privateroom',
 	secretroom: 'privateroom',
 	privateroom: function (target, room, user, connection, cmd) {
+		if (room.isPersonal) return this.errorReply("Personal rooms configuration can't be changed.");
```
Is this necessary? I detailed above why I think public and hidden are valid settings for personal rooms.

Joimer replied:
I'm changing this due to the changes mentioned above.

Zarel commented on commands.js:
```diff
@@ -366,7 +412,7 @@ var commands = exports.commands = {
 			if (room.chatRoomData) {
 				room.chatRoomData.modjoin = room.modjoin;
-				Rooms.global.writeChatRoomData();
+				if (!room.isPersonal) Rooms.global.writeChatRoomData();
```
As mentioned, this should not be here. If you don't want chat room data to be written, `room.chatRoomData` should not exist in the first place. Just don't create `.chatRoomData` for personal rooms.

Zarel replied:
This goes for every other instance of `Rooms.global.writeChatRoomData();`

Joimer replied:
Yup, done (on next PR update).

Zarel commented on commands.js:
```diff
@@ -400,6 +446,7 @@ var commands = exports.commands = {
 			return;
 		}
 		if (!this.can('declare')) return false;
+		if (room.isPersonal) return this.errorReply("Personal rooms configuration can't be changed.");
```
I'm not sure this is particularly necessary, either. I guess it can't hurt, though...

Joimer replied:
I'm removing this considering the other changes mentioned above.

Joimer commented:
Made the personal rooms to lack ``chatRoomData``, thus allowed for their customisation and also removed the checks for ``writeChatRoomData``.

Made ``room.modchat`` default to ``+`` along with modjoin, since the idea is to be treated like a private conversation where only invited people can join.

Joimer commented:
Indeed, so made it private again, that way it will be treated as a private conversation.

Zarel commented:
I'm not sure the default should be super-private, though.

We should probably make the choice available to the user:

- secret: only roomauth can see
- hidden: drivers/moderators can also see
- public: everyone can see

Joimer commented:
I'm adamant that the default option should be ``secret``, but it's nice to add a second parameter to specify the type of room.

Joimer commented:
Added a new parameter to choose room type, defaults to private.

Zarel commented on app.js:
```diff
@@ -205,6 +207,22 @@ global.ResourceMonitor = {
+		var duration = now - this.groupChatTimes[ip];
+		if (ip in this.groupChats && duration < 60 * 60 * 1000) {
+			this.groupChats[ip]++;
+			if (this.groupChats[ip] > 10) {
```
10 is a lot. Maybe 4?

Joimer replied:
Sounds right.

Zarel commented on commands.js:
```diff
@@ -511,6 +560,12 @@ var commands = exports.commands = {
+		if (room.isPersonal) {
+			if (!user.personalRooms) user.personalRooms = 0;
+			if (user.personalRooms + 1 >= MAX_PERSONAL_ROOMS) return this.sendReply("This user already has " + MAX_PERSONAL_ROOMS + " personal rooms.");
+			user.personalRooms++;
```
The user object can be flushed really easily. It's okay for caches, but not for anything intended to be persistent. I'd say remove this limitation entirely.

Zarel commented:
`/groupchat` doesn't clearly convey `create a group chat`. I think it should be just an informational command, and creating it should be in `/groupchat create` or something...

Zarel commented:
Especially since creation is rate-limited, accidentally creating one with `/groupchat` might be a bit worrying.

Joimer commented:
I disagree here.

If you are going to use the room anyways, or you are not going to use it, why care? I don't want to overcomplicate this as it is supposed to be easy and fast to use. A long name or many options aren't neither easy nor fast, people don't want to remember commands like ``/createagroupchatthat with, many, options, like, this``. They want to type a fast command, like IRC's ``/join name`` and be done with it.

The ``/makechatroom`` command and its friend ``/makeprivatechatroom`` are good examples, they may be clear but their verbosity is obnoxious. I'd say: leave that for Java.

MonoHearted commented:
I personally don't see any issue with `/groupchat new, x` – It's not even that many parameters, and it follows the syntax of PS's other commands.

Zarel commented:
I know, `/makechatroom` was intentionally made to be verbose and obnoxious, because the feature was in beta when it was introduced.

I guess my non-beta name for the command would be like `/chatroom create`, and for this to be analogously `/groupchat create`. `new` could be an alias.

Joimer commented:
Plus we already have ``/help groupchat``, which could work like ``/help room``. We should standarise help commands and make them be used just like that, I think.

Zarel commented:
The other issue is: How often do you expect someone to create a new room? I honestly think that `/groupchat create` is short enough considering how often it's used.

Joimer commented:
I don't think it makes sense for ``/groupchat`` to be an informative command, when there's ``/help groupchat`` (and possibly an updated `` /help room``?) and there's only going to be a single action, ``create``.

You'd expect more commands if you design a command in such a way, yet there's only one action: either read how to create a group chat, or create a group chat. The first purpose is served by the wonderful ``/help`` command, which right now is implemented in a clever and useful way. The second purpose is fulfilled by a single action command with optional parameters for name and room type, should you need them.

Why complicate things? KISS.

Zarel commented:
There can be more actions. `/groupchat delete`, `/groupchat list` (which should be the default action of `/groupchat` and is what I mean by "informational").

The complication isn't for no reason, it makes it clear what it's doing.

Joimer commented:
I still disagree.

I fail to see the use of a group chat list. It's not like you can list the private conversations of other users in any other software. I don't really know of any case, at least. And if you want to delete a room, roomban everyone and leave it. I don't think a feature that is meant for small groups of friends to have a private conversation really needs an instant kick+delete option.

Joimer commented:
I have made a quick survey. Most users don't want a feature like ``/groupchat delete`` even to exist.

Slayer95 commented:
Scratch commands... I'll only like the implementation if it fixes https://github.com/Zarel/Pokemon-Showdown-Client/issues/28.

But +1 for the feature.

Zarel commented:
It sounded to me that Starbloom at least thought `/groupchat create` is better syntax than `/groupchat`.

TheFenderStory commented:
@nineage wanted me to point out that we might want to add a check in the tournament create command. It looks like users can start tournaments in personal rooms right now, which isn't really in with Joim's idea of a small group of friends having a group chat, and could cause high memory usage if enough people use this feature.

Joimer commented:
Yes, a couple of users liked the ``create`` keyword, that's why I said "most" instead of "all of them". But if the opinion of a few people is more important than the rest of the userbase, well, I just want to get this feature done and available because it is highly expected by the userbase.

@TheFenderStory I considered it, I don't think it is an actual problem. If it becomes one, though, we can limit it

Zarel commented:
We already have `/tour create`, I'm planning on moving `/makechatroom` to something like `/chatroom create`, so I think it would be consistent with the rest of the code for this to be `/groupchat create`.

Joimer commented:
Ok, whatever.

Joimer commented:
I've added the ``create`` option, but also added a ``gcc`` alias that has an implicit create. That way both you and the userbase are happy and get what they want!

Zarel commented:
Does the userbase actually want a fast way to create a groupchat, and if so, why?

Zarel commented:
The entire idea behind doing it that way is because it should not be particularly easy to create a group chat, and it should especially not be easy to accidentally create a group chat. Both of which having the alias kind of defeats.

Joimer commented:
Yes they want. There are two main reasons:
If the come from Smogon Culture, they are used to IRC, when they just type ``/join name`` and they have their personal room created. This feature wants to adapt that to PS. The most consolidated Smogon userbase are expecting a short, easy to use, easy to remember command that let's them create a personal room instantly.

From the PS userbase, they usually talk about how they dislike complex or otherwise long commands, wanting to get things done easily. Since we lack buttons, short and easy commands is the way to go. Most people in development room specified they preferred to not to have to use an extra word just to create a group PM, besides a couple of users, most of them thought it unnecessary.

It should be easy to create a group chat. Every software that calls itself decent and has private messaging has a way to create very easily a group chat. As for the accidental creation thereof, the alias is going to be used just by those wanting to instantly create a room, as it is only known if you read the command help or someone else tells you "hey, /gcc name creates a room directly, awesome!".

Zarel commented:
Okay, we're clearly stalling here.

I'm going to take the same approach I did with public chatrooms.

Namely, we are going to name this ``/makegroupchat`` for now.

And after this is merged, we can decide on the finalized name at our leisure, and in the meantime the userbase can use ``/makegroupchat``.

Joimer commented:
Okay, updated accordingly.

Zarel commented on commands.js:
```diff
@@ -286,6 +286,48 @@ var commands = exports.commands = {
+		// Check if the custom part of the room title is empty. It needs a title.
+		var roomName = targets[0] || Math.floor(Math.random() * 100000000);
+		// Chosen name have the same limitations as regular rooms.
+		if (roomName.includes(',') || roomName.includes('|') || roomName.includes('[') || roomName.includes('-')) {
```
This crashes if `roomName` is an integer (i.e. if the command is called without any arguments)
